shared/types/indexer.go

```diff
@@ -0,0 +1,240 @@
+	// Index indexes by `(hash, height, sender, recipient)`
+	Index(result TxResult) error
+
+	// GetByHash returns all transaction specified by hash or nil if the transaction is not indexed
+	GetByHash(hash []byte) (TxResult, error)
+
+	// GetByHeight returns all transactions specified by height or nil if there are no transactions at that height
+	GetByHeight(height int64) ([]TxResult, error)
+
+	// GetBySender returns all transactions signed by *sender* may be ordered descending/ascending
+	GetBySender(sender string, descending bool) ([]TxResult, error)
+
+	// GetByRecipient returns all transactions *sent to address* may be ordered descending/ascending
+	GetByRecipient(rec string, descending bool) ([]TxResult, error)
+
+	// Close stops the underlying db connection
+	Close() error
+}
+
+type TxResult interface {
+	GetTx() []byte                        // the transaction object primitive
+	GetHeight() int64                     // height it was sent
+	GetIndex() int32                      // which index it was within the block-transactions; ordered by when the proposer received it in the mempool
+	GetResultCode() int32                 // 0 is no error, otherwise corresponds to error object code
+	GetError() string                     // can be empty
+	GetSigner() string                    // get the address who signed
+	GetRecipient() string                 // can be empty
+	GetMessageType() string               // corresponds to type of message (Ex. validator-stake, app-unjail, node-stake) etc.
+	Hash() ([]byte, error)                // the hash of the tx bytes
+	HashFromBytes([]byte) ([]byte, error) // the hash of the tx bytes, avoids re-marshalling
+	Bytes() ([]byte, error)               // proto marshalled bytes
+	FromBytes([]byte) (TxResult, error)   // from proto marshalled bytes
+}
+
+var _ TxResult = &DefaultTxResult{}
+var _ TxIndexer = &txIndexer{}
+
+// Implementation
+
+// txIndexer implementation uses a KVStore (interface) to index the transactions
+//
+// The transaction is indexed in the following formats:
+// - HASHKEY:      "h/SHA3(TxResultProtoBytes)"  VAL: TxResultProtoBytes     // store value by hash
+// - HEIGHTKEY:    "b/height/index"              VAL: HASHKEY                // store hashKey by height
+// - SENDERKEY:    "s/height/index"              VAL: HASHKEY                // store hashKey by sender
+// - RECIPIENTKEY: "r/height/index"              VAL: HASHKEY                // store hashKey by recipient (if not empty)
+//
+// FOOTNOTE: the height/index is store using [ELEN](https://github.com/jordanorelli/lexnum/blob/master/elen.pdf)
+// This is to ensure the results are stored sorted (assuming KVStore uses a byte-wise lexicographical sorting)
+//
+
+const (
+	HashPrefix            = 'h'
+	HeightPrefix          = 'b' // b for block
+	SenderPrefix          = 's'
+	RecipientPrefix       = 'r'
+	DefaultHeightOrdering = true
+)
+
+// =,- are the default parameters in the example repository.
```

Comment:
Please add a reference link

Reply:
I don't think this is needed because it's included in the source code dependency README.md